### PR TITLE
TY: Support type inference for GATs

### PR DESCRIPTION
**Pull request by mchernyavsky**
Relates to https://github.com/intellij-rust/intellij-rust/issues/6633.

Depends on https://github.com/intellij-rust/intellij-rust/pull/8542.

changelog: Provide initial type inference support for [GATs](https://blog.rust-lang.org/2022/10/28/gats-stabilization.html)

### CODE REVIEW

File: src/test/kotlin/org/rust/lang/core/type/RsGenericExpressionTypeInferenceTest.kt [hunk 1]
```diff
@@ -1742,6 +2021,20 @@ class RsGenericExpressionTypeInferenceTest : RsTypificationTestBase() {
...
+        impl Trait for S {
+            type Item<U> = X<U>;
+        }
+        fn main() {
+            let foo1 = Foo::<S, u8>(X);
+
+            foo1.0;
+        }      //^ X<u8>
+    """)
+
```
Comment (vlad20012) [inline src/test/kotlin/org/rust/lang/core/type/RsGenericExpressionTypeInferenceTest.kt:2037]: Could you please add tests with const parameters as well?

**Review by vlad20012 — APPROVED** [review]

File: src/main/kotlin/org/rust/ide/hints/type/RsTypeHintsPresentationFactory.kt [hunk 2]
```diff
@@ -121,6 +121,27 @@ class RsTypeHintsPresentationFactory(
...
+            suffix = text(">"),
+            startWithPlaceholder = checkSize(level, 2)
+        )
+
+        val target = type.target.element
+        val targetName = target.name
+        val targetPresentation = factory.psiSingleReference(text(targetName)) { target }
+        val typeNamePresentation = listOf(collapsible, targetPresentation).join("::")
+        val typeArguments = type.typeArguments.zip(target.typeParameters)
+        return withGenericsTypeHint(typeNamePresentation, typeArguments, emptyList(), level)
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply src/main/kotlin/org/rust/ide/hints/type/RsTypeHintsPresentationFactory.kt:142]: I don't have the required information in `traitItemTypeHint`. I need to change `assoc` type in `BoundElement` from `Map<RsTypeAlias, Ty>` to `Map<BoundElement<RsTypeAlias>, Ty>` in order to fix it.

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt [hunk 3]
```diff
@@ -1136,12 +1137,12 @@ class ImplLookup(
         vararg subst: Ty
     ): SelectionResult<TyWithObligations<Ty>?> {
         val (trait, assocType) = traitAndOutput
-        return selectProjection(TraitRef(selfTy, trait.withSubst(*subst)), assocType)
+        return selectProjection(TraitRef(selfTy, trait.withSubst(*subst)), assocType.withSubst(*subst))
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt:1140]:
```suggestion
        return selectProjection(TraitRef(selfTy, trait.withSubst(*subst)), assocType.withSubst())
```

**Review by vlad20012 — APPROVED** [review]
bors r+

**Pull request merged**